File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void letterCombinationsHelper(int j,string digits,string& temp,vector<string>& n,vector<string>&ans){
        if(j==digits.size()) { 
            ans.push_back(temp);
            return;
        }
        int num = digits[j]-'0';
        cout<<num<<endl;
        for(char c : n[num]){
             temp = temp+c;
             letterCombinationsHelper(j+1,digits,temp,n,ans);
             temp.pop_back();
        } 
    }
    vector<string> letterCombinations(string digits) {
      vector<string> n = {"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
      vector<string> ans;
      string temp ="";
      if(digits=="") return ans;
      letterCombinationsHelper(0,digits,temp,n,ans);
     return ans;
      }
      
    };

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        if(digits.length()==0) return ans;
        String[] map = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<Character> temp = new ArrayList<>();
        Helper(0,digits,map,temp,ans);
        return ans;
    }
    public void Helper(int digit, String digits,String[] map,List<Character> temp,List<String> ans)
    {
        if(digit==digits.length()){
            StringBuilder sb = new StringBuilder();
            for(char c: temp){
                sb.append(c);
            }
            ans.add(sb.toString());
            return;
        }
        int num = digits.charAt(digit)-'0';
        for( char c : map[num].toCharArray())
        {
            temp.add(c);
            Helper(digit+1,digits,map,temp,ans);
            temp.remove(temp.size()-1);
        }
    }
}